keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,552 @@
+- Pods e2e test: https://github.com/kubernetes/kubernetes/blob/91ee30074bee617d52fc24dc85132fe948aa5153/test/e2e/common/node/pods.go
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag.
+- unit and integration tests completed.
+- Initial e2e tests completed and enabled.
+
+#### Beta
+
+- Feature gate is enabled by default.
+- Allowing time for feedback.
+
+#### GA
+
+- TBD
+
+### Upgrade / Downgrade Strategy
+
+kube-apiserver should be updated before kubelets in that order. Upgrade involves
+draining all pods from the node, update to a matching kubelet and making the
+node schedulable again. Downgrade involves doing the above in reverse.
+
+### Version Skew Strategy
+
+Since kubelet versions must not be ahead of kube-apiserver versions, the older
+versions of kubelet or if the feature gate is disabled in kubelet, the fields
+will be ignored.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: EnvFiles
+  - Components depending on the feature gate:
+    - kube-apiserver
+    - kubelet
+
+###### Does enabling the feature change any default behavior?
+
+No
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, Containers already running will have their environment variables populated
+from the specified file. But on restart, the fields will be stripped and the
+feature will no longer work.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Newly started or restarted containers will have their environment variables
+populated from the specified file.
+
+###### Are there any tests for feature enablement/disablement?
+
+No
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+The [Version Skew Strategy](#version-skew-strategy) section covers this point.
+It will not impact any already running workloads.
+
+
+###### What specific metrics should inform a rollback?
+
+An increase in pod validation errors can indicate issues with the field
+translation. These would show up as `code=400` (Bad Request) errors in
+`apiserver_request_total`.
+
+The following errors could indicate problems with how kubelets are working
+with this new fields:
+
+- `started_containers_errors_total`
+- `started_pods_errors_total`
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+Automated tests will cover the scenarios with and without the changes proposed
+on this KEP. As defined under [Version Skew Strategy](#version-skew-strategy),
+we are assuming the cluster may have kubelets with older versions (without
+this KEP' changes), therefore this will be covered as part of the new tests.
```

Comment:
Let's list out the key tests we need to ensure the upgrade->downgrade->upgrade is covered here. E.g.

- When the kubelet is downgraded, the env files are no used to provide envs to containers, but the pod otherwise runs
- When the apiserver is downgraded, new envs cannot be written but existing envs continue to exist and can be removed, when upgraded, the envs can again be written..
- ??